docs/proposals/CSV data mapping.md

```diff
@@ -0,0 +1,122 @@
+# Proposal: Introduce APIs to perform conversions between the Map/Ballerina record and XML
+
+_Owners_: @daneshk @sahanhe  
+_Reviewers_: @daneshk  
+_Created_: 2022/07/04   
+_Updated_: -  
+_Issues_: [#3004](https://github.com/ballerina-platform/ballerina-standard-library/issues/3004#event-6837345074)
+
+
+## Summary
+
+Current CSV read/write APIs supports reading and writing as string[][] values. In this proposal we propose to add data mapping to return an array of particular record defined by the user.
+
+## Goals
+
+* Adding data mapping feature to CSV READ/Write operations to return an array of particular record defined by the user.
+
+* Change ```fileReadCsv, fileWriteCsv, fileReadCsvAsStream, fileWriteCsvFromStream``` to support data mapping.
+
+* Carry out the data mapping by inferring the return type automatically as a dependently-typed function.
+
+## Motivation
+
+Until now, user doesn't have a direct data mapping method to get an array of user defined records from CSVs.
+
+
+
+## Description
+
+```fileReadCsv(), fileWriteCsv(), fileReadCsvAsStream(), fileWriteCsvFromStream()```  APIs are modified to support data mapping in CSV read/write operations. Existing Currently ```fileReadCsv()``` API returns String[][] and modified API can return either String[][] or record[] depending on the return type. Similarly ```fileWriteCsv()``` currently accepts String[][] inputs and modified API can accept both String[][] and record[].
+
+For stream type APIs, ```fileReadCsvAsStream()``` currently returns ```stream<string[], Error?>|Error``` and modified API will return ```stream<string[]|record{}, Error?>|Error```.  Similarly, ```fileWriteCsvFromStream()``` accepts ```stream<string[], Error?>```  as the input and it will be modified to ```stream<string[]|record{}, Error?>```.
```

Comment:
```suggestion
* The ```fileReadCsvAsStream()``` currently returns ```stream<string[], Error?>|Error``` and it will be modified to return ```stream<string[]|record{}, Error?>|Error```.  
* The ```fileWriteCsvFromStream()``` currently accepts ```stream<string[], Error?>```  as the input and it will be modified to accept ```stream<string[]|record{}, Error?>```.
```